pkg/bootstrap/static.go

```diff
@@ -0,0 +1,392 @@
+			Hosts:                []*core.Address{telemetryAddress},
+			CircuitBreakers:      circuitBreakers,
+			Http2ProtocolOptions: &core.Http2ProtocolOptions{},
+		}
+		if opts.ControlPlaneAuth {
+			cluster.TlsContext = &auth.UpstreamTlsContext{
+				CommonTlsContext: commonTLSContext([]string{opts.TelemetrySAN}),
+			}
+		}
+		out.StaticResources.Clusters = append(out.StaticResources.Clusters, cluster)
+	} else {
+		telemetryCluster = clusterName(*opts.TelemetryUpstream)
+	}
+
+	out.StaticResources.Listeners = BuildListeners(opts.Upstreams, telemetryCluster, opts.ControlPlaneAuth)
+
+	if err := out.Validate(); err != nil {
+		return nil, err
+	}
+
+	return &out, nil
+}
+
+func commonTLSContext(altNames []string) *auth.CommonTlsContext {
+	return &auth.CommonTlsContext{
+		TlsCertificates: []*auth.TlsCertificate{{
+			CertificateChain: &core.DataSource{
+				Specifier: &core.DataSource_Filename{Filename: "/etc/certs/cert-chain.pem"},
+			},
+			PrivateKey: &core.DataSource{
+				Specifier: &core.DataSource_Filename{Filename: "/etc/certs/key.pem"},
+			},
+		}},
+		ValidationContext: &auth.CertificateValidationContext{
+			TrustedCa: &core.DataSource{
+				Specifier: &core.DataSource_Filename{Filename: "/etc/certs/root-cert.pem"},
+			},
+			VerifySubjectAltName: altNames,
+		},
+	}
+}
+
+func toAddress(address string) (*core.Address, error) {
+	host, portStr, err := net.SplitHostPort(address)
+	if err != nil {
+		return nil, err
+	}
+	port, err := strconv.Atoi(portStr)
+	if err != nil {
+		return nil, err
+	}
+	return &core.Address{
+		Address: &core.Address_SocketAddress{
+			SocketAddress: &core.SocketAddress{
+				Address:       host,
+				PortSpecifier: &core.SocketAddress_PortValue{PortValue: uint32(port)},
+			},
+		},
+	}, nil
+}
+
+func toStruct(msg proto.Message) *types.Struct {
+	out, err := util.MessageToStruct(msg)
+	if err != nil {
+		panic(err)
+	}
+	return out
+}
+
+// BuildServiceConfig generates a mixer client config.
+func BuildServiceConfig(service, uid, telemetryCluster string) *mixerclient.HttpClientConfig {
+	return &mixerclient.HttpClientConfig{
+		DefaultDestinationService: service,
+		ServiceConfigs: map[string]*mixerclient.ServiceConfig{
+			service: {
+				DisableCheckCalls:  true,
+				DisableReportCalls: false,
+				MixerAttributes: &mixerv1.Attributes{
+					Attributes: map[string]*mixerv1.Attributes_AttributeValue{
+						"destination.service": {Value: &mixerv1.Attributes_AttributeValue_StringValue{
+							StringValue: service,
+						}},
+						"destination.uid": {Value: &mixerv1.Attributes_AttributeValue_StringValue{
+							StringValue: uid,
+						}},
+					},
+				},
+			},
+		},
+		Transport: &mixerclient.TransportConfig{
+			CheckCluster:  "mixer_check_server",
+			ReportCluster: telemetryCluster,
+		},
+	}
+}
+
+// BuildListeners generates HTTP listeners from upstreams
+func BuildListeners(upstreams []Upstream, telemetryCluster string, controlPlaneAuth bool) []v2.Listener {
```

Comment:
Please reuse existing listener builders and don't just do it all here. We want single consolidated locations to produce this configuration so we have one place to fix problems, rather than having bespoke stuff spread all over the codebase.

Reply:
Because the whole goal is to consolidate the logic.

Reply:
No, that's not the goal. The goal is to get telemetry for istio components. How it's done is not consequential. I can clean it up later, once the code settles.